### PR TITLE
[IPU] support fleet api

### PR DESCRIPTION
**Pull request by gglin001**
### PR types

Others
### PR changes

Others
### Describe

support fleet using poprun

### CODE REVIEW

**Review by qili93 — DISMISSED** [review]
LGTM

File: python/paddle/distributed/fleet/launch.py [hunk 1]
```diff
@@ -432,6 +459,78 @@ def launch_ps(args, distribute_mode):
...
+def launch_ipu(args):
+    # The number of replicas for data parallel
+    assert (args.num_ipus % args.ipus_per_replica) == 0, \
+                "The number of IPUs:{} mod the number of IPUs per replica:{} must == 0".format(args.num_ipus, args.ipus_per_replica)
+    num_replicas = int(args.num_ipus / args.ipus_per_replica)
+
+    # The number of processes
+    nproc_per_node = 1
+    if args.nproc_per_node is not None:
+        nproc_per_node = args.nproc_per_node
```
Comment (sneaxiy) [inline python/paddle/distributed/fleet/launch.py:471]: What is the difference between `nproc_per_node` and `ipus_per_replica`?

File: python/paddle/distributed/fleet/launch.py [hunk 2]
```diff
@@ -432,6 +459,78 @@ def launch_ps(args, distribute_mode):
...
+    # local envs
+    for idx in range(num_procs):
+        cur_endpoint = endpoints[idx // nproc_per_node]
+        rank_in_node = idx % nproc_per_node
+        poprun_args.append(
+            '--instance-mpi-local-args={}:\"-x PADDLE_TRAINER_ID={} -x PADDLE_CURRENT_ENDPOINT={} -x PADDLE_RANK_IN_NODE={}\"'.
+            format(idx, idx, cur_endpoint, rank_in_node))
+
+    # executor
+    poprun_args.append("python3.7")
```
Comment (sneaxiy) [inline python/paddle/distributed/fleet/launch.py]: How about `poprun_args.append(sys.executable)`?

**Comment by gglin001** [comment]
@yaozhixin 👆

File: python/paddle/distributed/fleet/launch.py (hunk 2 above)
Comment (yaozhixin) [reply python/paddle/distributed/fleet/launch.py]: ok

File: python/paddle/distributed/fleet/launch.py (hunk 1 above)
Comment (yaozhixin) [reply python/paddle/distributed/fleet/launch.py:471]: For poprun distributed computing, we have the following concepts:
1. ipus_per_replica: the number of IPUs per instance.

2. replica:  data parallel working in a process. For example: Require 2 IPUs (ipus_per_replica=2) per instance, and set replicas=2 (2 instances), then PopART will totally acquire 4 IPUs per process for data parallel.

3. process: each process is able to include multi instances (replicas > 1)

4. hosts (ips) : each host is able to include multi processes.

`nproc_per_node`: The number of processes per host.
`ipus_per_replica`: The number of IPUs per instance (per replica).

File: python/paddle/distributed/fleet/launch.py (hunk 2 above)
Comment (gglin001) [reply python/paddle/distributed/fleet/launch.py]: fixed

**Review by sneaxiy — APPROVED** [review]
LGTM.

**Comment by sneaxiy** [comment]
@kuizhiqing Please help to review the parameters in `launch.py`.

**Comment by gglin001** [comment]
will update this part on the new paddle distributed API

**Pull request closed without merging**